src/api/app/models/package.rb

```diff
@@ -1516,10 +1519,10 @@ def update_activity
     # the value we add to the activity, when the object gets updated
     addon = 10 * (Time.now.to_f - updated_at_was.to_f) / 86400
     addon = 10 if addon > 10
-    new_activity = activity + addon
-    new_activity > 100 ? 100 : new_activity
 
-    self.activity_index = new_activity
+    # atomar update to avoid dead locks
+    sql = ApplicationRecord.connection
+    sql.execute("UPDATE packages SET activity_index = LEAST(100, #{Package._activity_algorithm} + #{addon})  WHERE id=#{id}")
```

Comment:
Why is this raw sql? I assume because otherwise you end up in a loop because of the before_update hook?

Reply:
There are other way to [skip callbacks](http://guides.rubyonrails.org/active_record_callbacks.html#skipping-callbacks) if that is really the reason you try this...

Reply:
activity was sql code before.

Feel free  to change this if there is away.

Reply:
Why not getting rid of the SQL at all! We know the activity_index already and calculating the unix timestamp and power we can do in ruby. I doubt that the database is faster in math calculations here.

```
self.activity_index = activity_index * (2.3276 ** ((updated_at.to_f - Time.now.to_f) / 10000000))
```

Reply:
No need to convert this to floats. Just run `updated_at - Time.now`.

Reply:
btw, you would have to use the minimum of 100 and the result of your expression, to have the same behaviour like before.

Reply:
@bgeuken if you leave the code like before and just replace the SQL it is already covered because it was already in the ruby code before.

Yeah, subtracting Time works, however, the precision is higher when parsing to_f before! 
Actually, to have a more exact result you would need to do:
```
(2.3276 ** ((updated_at.to_f.round(6) - Time.now.to_i) / 10000000))
```

Reply:
The database can be way faster when doing this for many packages, but I am also fine changing this. My goal here was just to solve the dead lock crashes in first place and trying change as little as possible there.

Reply:
@bgeuken LEAST(100, ...) is limiting it to value 100 like before, no?

Reply:
I meant [Christian's proposal](https://github.com/openSUSE/open-build-service/pull/4171#discussion_r154071424).

Reply:
> ChrisBr 4 days ago  Member
> @bgeuken if you leave the code like before and just replace the SQL it is already covered because it was already in the ruby code before.

Ah, alright.